src/components/PayPalScriptProvider.tsx

```diff
@@ -0,0 +1,73 @@
+import React from "react";
+import { loadScript } from "@paypal/paypal-js";
+
+import {
+    ScriptProviderProps,
+    SCRIPT_LOADING_STATE,
+    DISPATCH_ACTION,
+} from "../types/ScriptProvider";
+import {
+    getScriptID,
+    ScriptContext,
+    scriptReducer,
+} from "../context/ScriptProvider";
+import { SCRIPT_ID } from "../constants";
+
+export const PayPalScriptProvider: React.FC<ScriptProviderProps> = ({
+    options,
+    children,
+    deferLoading = false,
+}: ScriptProviderProps) => {
+    const [state, dispatch] = React.useReducer(scriptReducer, {
+        options: {
+            ...options,
+            [SCRIPT_ID]: `${getScriptID(options)}`,
+        },
+        loadingStatus: deferLoading
+            ? SCRIPT_LOADING_STATE.INITIAL
+            : SCRIPT_LOADING_STATE.PENDING,
+        dispatch: null,
+    });
+
+    React.useEffect(() => {
+        if (
+            deferLoading === false &&
+            state.loadingStatus === SCRIPT_LOADING_STATE.INITIAL
+        ) {
+            return dispatch({
+                type: DISPATCH_ACTION.LOADING_STATUS,
+                value: SCRIPT_LOADING_STATE.PENDING,
+            });
+        }
+
+        if (state.loadingStatus !== SCRIPT_LOADING_STATE.PENDING) return;
+
+        let isSubscribed = true;
+        loadScript(state.options)
+            .then(() => {
+                if (isSubscribed) {
+                    dispatch({
+                        type: DISPATCH_ACTION.LOADING_STATUS,
+                        value: SCRIPT_LOADING_STATE.RESOLVED,
+                    });
+                }
+            })
+            .catch(() => {
+                if (isSubscribed) {
+                    dispatch({
+                        type: DISPATCH_ACTION.LOADING_STATUS,
+                        value: SCRIPT_LOADING_STATE.REJECTED,
+                    });
+                }
+            });
+        return () => {
+            isSubscribed = false;
+        };
+    }, [options, deferLoading, state.loadingStatus]);
```

Comment:
In all the `useEffect` calls, we are purposely excluding some dependencies to avoid re-rendering. Let's add in this comment to each `useEffect` call so eslint does not warn about this.
```suggestion
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [options, deferLoading, state.loadingStatus]);
```

Reply:
Here I can change the dependency to state.options if you agree, inside the useEffect we aren't using the options object we received in the props of the component

Reply:
Ahh good point. Let’s update this one